### PR TITLE
Fix JWE encryption with user provided secret key

### PR DESCRIPTION
**Pull request by wendigo**
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Review by s2lomon — APPROVED** [review]
I like this changes, but I was wondering whether we haven't decided that we prefer to not allow using weaker keys, where your pr adds support for this. Apart from that - lgtm.

File: core/trino-main/src/main/java/io/trino/server/security/oauth2/TokenPairSerializer.java [hunk 1]
```diff
@@ -88,5 +89,34 @@ public Optional<String> getRefreshToken()
         {
             return refreshToken;
         }
+
+        @Override
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/server/security/oauth2/TokenPairSerializer.java:93]: Should TokenPair be a record?

Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/server/security/oauth2/TokenPairSerializer.java]: Converted to a record

File: core/trino-main/src/test/java/io/trino/server/security/oauth2/TestJweTokenSerializer.java [hunk 2]
```diff
@@ -45,15 +48,87 @@ public class TestJweTokenSerializer
...
-        assertThat(deserializedTokenPair.getAccessToken()).isEqualTo("access_token");
-        assertThat(deserializedTokenPair.getExpiration()).isEqualTo(expiration);
-        assertThat(deserializedTokenPair.getRefreshToken()).isEqualTo(Optional.of("refresh_token"));
+        assertThat(deserializedTokenPair).isEqualTo(tokenPair);
+        assertThat(deserializedTokenPair.accessToken()).isEqualTo("access-token");
+        assertThat(deserializedTokenPair.refreshToken()).isEqualTo(Optional.of("refresh-token"));
+    }
+
+    @Test
+    public void testDeserializationWithWrongSecret()
```
Comment (lukasz-walkiewicz) [inline core/trino-main/src/test/java/io/trino/server/security/oauth2/TestJweTokenSerializer.java:64]: Parametrized test?

File: core/trino-main/src/test/java/io/trino/server/security/oauth2/TestJweTokenSerializer.java [hunk 3]
```diff
@@ -45,15 +48,87 @@ public class TestJweTokenSerializer
...
+
+        // This will use wrong secret keys to decrypt
+        assertThatThrownBy(() -> assertRoundTrip(Optional.empty(), Optional.of(randomEncodedSecret())))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("Decryption failed")
+                .hasStackTraceContaining("Integrity check failed");
+    }
+
+    @Test
+    public void testSerializationDeserializationRoundTripWithDifferentKeyLengths()
```
Comment (lukasz-walkiewicz) [inline core/trino-main/src/test/java/io/trino/server/security/oauth2/TestJweTokenSerializer.java:90]: Parametrized test?

File: core/trino-main/src/test/java/io/trino/server/security/oauth2/TestJweTokenSerializer.java [hunk 4]
```diff
@@ -181,4 +266,17 @@ public void advanceBy(Duration currentTimeDelta)
             this.currentTime = currentTime.plus(currentTimeDelta.toMillis(), MILLIS);
         }
     }
+
+    private static String randomEncodedSecret()
+    {
+        return randomEncodedSecret(24);
```
Comment (lukasz-walkiewicz) [inline core/trino-main/src/test/java/io/trino/server/security/oauth2/TestJweTokenSerializer.java:272]: 32?

File: core/trino-main/src/main/java/io/trino/server/security/oauth2/JweTokenSerializer.java [hunk 5]
```diff
@@ -153,6 +157,28 @@ public String serialize(TokenPair tokenPair)
         }
     }
 
+    private JWEAlgorithm getEncryptionAlgorithm(SecretKey key)
```
Comment (lukasz-walkiewicz) [inline core/trino-main/src/main/java/io/trino/server/security/oauth2/JweTokenSerializer.java:160]: How about a single method which returns `JWEHeader`?

File: core/trino-main/src/test/java/io/trino/server/security/oauth2/TestJweTokenSerializer.java (hunk 4 above)
Comment (wendigo) [reply core/trino-main/src/test/java/io/trino/server/security/oauth2/TestJweTokenSerializer.java:272]: 24 is allowed value. Doesn't matter

File: core/trino-main/src/main/java/io/trino/server/security/oauth2/JweTokenSerializer.java (hunk 5 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/server/security/oauth2/JweTokenSerializer.java:160]: Makes sense

File: core/trino-main/src/main/java/io/trino/server/security/oauth2/JweTokenSerializer.java [hunk 6]
```diff
@@ -50,8 +52,7 @@ public class JweTokenSerializer
         implements TokenPairSerializer
 {
     private static final Logger LOG = Logger.get(JweTokenSerializer.class);
-    private static final JWEAlgorithm ALGORITHM = JWEAlgorithm.A256KW;
-    private static final EncryptionMethod ENCRYPTION_METHOD = EncryptionMethod.A256CBC_HS512;
+    private final JWEHeader encryptionHeader;
```
Comment (lukasz-walkiewicz) [inline core/trino-main/src/main/java/io/trino/server/security/oauth2/JweTokenSerializer.java]: nit: move below static fields

**Review by lukasz-walkiewicz — APPROVED** [review]

**Review by kokosing — COMMENTED** [review]
Refactors should be before actual change. That way you don't have add a thing (hashCode/equals) in one commit and then in the second commit remove it.

**Comment by wendigo** [comment]
@kokosing sure, I'll reorder commits

**Review by kokosing — APPROVED** [review]

**Comment by colebow** [comment]
Does this need a release node? @wendigo @kokosing

**Comment by wendigo** [comment]
@colebow yes. It fixes the ability to provide secret that’s used as symmetric key for JWE token encryption. 16, 24, 32 bytes keys are supported (which translates to 128, 178 and 256 bit encryption of the JWE token)

**Comment by wendigo** [comment]
User provided secret is important because on every cluster restart new key will be autogenerated when user is not providing it’s own. Then all of the sessions will be invalidated.

**Comment by sajjoseph** [comment]
@wendigo  - we came across an issue with current implementation.
When access token and refresh token were combined and written into a cookie, the size of the cookie (4334 bits) was larger than normal (4096 bits). Browsers (Chrome for example) were rejecting them. We got it working by writing these to separate cookies though. 

Has anyone come across that issue? Is it a good idea to separate out these tokens into their own cookies?

**Comment by wendigo** [comment]
@sajjoseph This won't be a backward compatible change. Maybe splitting them across 4096 ranges is a solution? Adding more cookies and combining them when reading?

**Comment by wendigo** [comment]
@sajjoseph https://github.com/trinodb/trino/pull/20787

**Comment by sajjoseph** [comment]
Awesome! Thanks. This will help us to use the OAuth Refresh token flow. I will try it out and let you know the results.

**Pull request merged**